File: src/ros2_bridge/mappers/nav_msgs/trajectory_point.rs
//! Typed mapper for `nav_msgs/msg/TrajectoryPoint`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn trajectory_point_to_bus(msg: ros_env::nav_msgs::msg::TrajectoryPoint) -> crate::nav_msgs::msg::v1::TrajectoryPoint {
    crate::nav_msgs::msg::v1::TrajectoryPoint {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        pose: Some(crate::ros2_bridge::mappers::geometry_msgs::pose::pose_to_bus(msg.pose)),
        velocity: Some(crate::ros2_bridge::mappers::geometry_msgs::twist::twist_to_bus(msg.velocity)),
        acceleration: Some(crate::ros2_bridge::mappers::geometry_msgs::accel::accel_to_bus(msg.acceleration)),
        effort: Some(crate::ros2_bridge::mappers::geometry_msgs::wrench::wrench_to_bus(msg.effort)),
    }
}

pub(crate) fn trajectory_point_to_ros(bus: crate::nav_msgs::msg::v1::TrajectoryPoint) -> ros_env::nav_msgs::msg::TrajectoryPoint {
    ros_env::nav_msgs::msg::TrajectoryPoint {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(bus.header.unwrap_or_default()),
        pose: crate::ros2_bridge::mappers::geometry_msgs::pose::pose_to_ros(bus.pose.unwrap_or_default()),
        velocity: crate::ros2_bridge::mappers::geometry_msgs::twist::twist_to_ros(bus.velocity.unwrap_or_default()),
        acceleration: crate::ros2_bridge::mappers::geometry_msgs::accel::accel_to_ros(bus.acceleration.unwrap_or_default()),
        effort: crate::ros2_bridge::mappers::geometry_msgs::wrench::wrench_to_ros(bus.effort.unwrap_or_default()),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct NavMsgsTrajectoryPointMapper;

impl TypedTopicMapper for NavMsgsTrajectoryPointMapper {
    type Ros = ros_env::nav_msgs::msg::TrajectoryPoint;
    type Bus = crate::nav_msgs::msg::v1::TrajectoryPoint;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(trajectory_point_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(trajectory_point_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/nav_msgs/trajectory_point.py
"""Generated mapper for `nav_msgs/msg/TrajectoryPoint`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.pose import pose_to_bus, pose_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.twist import twist_to_bus, twist_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.accel import accel_to_bus, accel_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.wrench import wrench_to_bus, wrench_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.nav_msgs.msg.v1 import TrajectoryPoint as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def trajectory_point_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.pose.CopyFrom(pose_to_bus(msg.pose))
    bus.velocity.CopyFrom(twist_to_bus(msg.velocity))
    bus.acceleration.CopyFrom(accel_to_bus(msg.acceleration))
    bus.effort.CopyFrom(wrench_to_bus(msg.effort))
    return bus


def trajectory_point_to_ros(bus):
    from nav_msgs.msg import TrajectoryPoint as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.pose = pose_to_ros(bus.pose)
    out.velocity = twist_to_ros(bus.velocity)
    out.acceleration = accel_to_ros(bus.acceleration)
    out.effort = wrench_to_ros(bus.effort)
    return out


class NavMsgsTrajectoryPointMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from nav_msgs.msg import TrajectoryPoint as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return trajectory_point_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return trajectory_point_to_ros(bus)
